### PR TITLE
coprocessor/dag/expr: add some un-pushed builtin UDFs (#2918)

### PR DESCRIPTION
**Pull request by bb7133**
Implemented more builtin UDFs for coprocessor

### CODE REVIEW

**Comment by ngaut** [comment]
Well done. Please fix circlec.
You may need to run "make format" and run tests before creating a PR

File: src/coprocessor/dag/expr/builtin_op.rs [hunk 1]
```diff
@@ -177,6 +177,39 @@ impl FnCall {
         let lhs = try_opt!(self.children[0].eval_int(ctx, row));
         Ok(Some(!lhs))
     }
+
+    pub fn bit_count(&self, ctx: &mut EvalContext, row: &[Datum]) -> Result<Option<i64>> {
+        let lhs = try_opt!(self.children[0].eval_int(ctx, row));
+        let mut cnt = 0;
+        let mut remains = lhs as u64;
+        while remains != 0 {
```
Comment (overvenus) [inline src/coprocessor/dag/expr/builtin_op.rs]: How about `u64::count_ones()`?

Comment (bb7133) [reply src/coprocessor/dag/expr/builtin_op.rs]: thanks for the reviews!

Comment (overvenus) [reply src/coprocessor/dag/expr/builtin_op.rs]: Thank you for the PR. :smile:

File: src/coprocessor/dag/expr/arithmetic.rs [hunk 2]
```diff
@@ -197,11 +197,128 @@ impl FnCall {
...
+        let lus = mysql::has_unsigned_flag(self.children[0].get_tp().get_flag());
+        let rus = mysql::has_unsigned_flag(self.children[1].get_tp().get_flag());
+
+        let data_type = if lus || rus {
+            "BIGINT UNSIGNED"
+        } else {
+            "BIGINT"
+        };
+
+        let overflow = Error::overflow(data_type, &format!("({} DIV {})", lhs, rhs));
```
[the remark this answers is not in the document]
Comment (bb7133) [reply src/coprocessor/dag/expr/arithmetic.rs]: I see, let me fix this, thanks!

**Comment by siddontang** [comment]
PTAL @AndreMouche @breeswish

File: src/coprocessor/dag/expr/arithmetic.rs [hunk 3]
```diff
@@ -197,11 +197,121 @@ impl FnCall {
...
+        if rhs == 0 {
+            return Ok(None);
+        }
+
+        let lus = mysql::has_unsigned_flag(self.children[0].get_tp().get_flag());
+        let rus = mysql::has_unsigned_flag(self.children[1].get_tp().get_flag());
+
+        let res = match (lus, rus) {
+            (true, true) => Some(((lhs as u64) / (rhs as u64)) as i64),
+            (false, false) => if lhs == i64::MIN && rhs == -1 {
```
Comment (breeswish) [inline src/coprocessor/dag/expr/arithmetic.rs]: Although many of these code are only occurred once, I suggest to extract overflow behaviors into a new module, exactly like what we did in TiDB, to reduce the effort of maintain in future and to make it easier to compare differences between TiDB's implementation and TiKV's implementation.

File: src/coprocessor/dag/expr/arithmetic.rs [hunk 4]
```diff
@@ -197,11 +197,121 @@ impl FnCall {
...
+                None
+            } else {
+                Some((lhs / rhs) as i64)
```
Comment (breeswish) [inline src/coprocessor/dag/expr/arithmetic.rs]: is `as i64` redundant?

File: src/coprocessor/dag/expr/arithmetic.rs [hunk 5]
```diff
@@ -197,11 +197,121 @@ impl FnCall {
...
+            },
+            (false, true) => if lhs < 0 {
+                if lhs.overflowing_neg().0 as u64 >= rhs as u64 {
+                    None
+                } else {
+                    Some(0 as i64)
+                }
+            } else {
+                Some(((lhs as u64) / rhs as u64) as i64)
```
Comment (breeswish) [inline src/coprocessor/dag/expr/arithmetic.rs]: `(rhs as u64)`

**Comment by huachaohuang** [comment]
This is a huge PR, and the functions are not related to each other, so I think we can split them into small PRs.

**Comment by bb7133** [comment]
Fine, I will split this PR into several pieces

**Comment by bb7133** [comment]
PTAL @huachaohuang @AndreMouche @breeswish, now this PR is focused on arithmetic.rs, I also ported overflow.go from TiDB with similar functionality, thanks.

File: src/coprocessor/codec/overflow.rs [hunk 6]
```diff
@@ -0,0 +1,331 @@
...
+// sub_i64 subtracts i64 a with b and returns i64 if no overflow error.
+pub fn sub_i64(a: i64, b: i64) -> Result<i64> {
+    a.checked_sub(b)
+        .ok_or_else(|| Error::overflow("BIGINT", &format!("({} - {})", a, b)))
+}
+
+// sub_u64_with_i64 subtracts u64 a with i64 b and returns u64 if no overflow error.
+pub fn sub_u64_with_i64(a: u64, b: i64) -> Result<u64> {
+    if b < 0 {
+        add_u64(a, b.overflowing_neg().0 as u64)
```
[the remark this answers is not in the document]
Comment (bb7133) [reply src/coprocessor/codec/overflow.rs]: Thanks, I will add the case in the future, but for now it will be removed from this PR

**Comment by bb7133** [comment]
Removed unused add_xxx/sub_xxx/mul_xxx functions in overflow.rs

File: src/coprocessor/dag/expr/arithmetic.rs [hunk 7]
```diff
@@ -197,11 +197,94 @@ impl FnCall {
             None => Ok(None),
         }
     }
+
+    pub fn intdivide_int(&self, ctx: &mut EvalContext, row: &[Datum]) -> Result<Option<i64>> {
```
Comment (AndreMouche) [inline src/coprocessor/dag/expr/arithmetic.rs]: int_divide_int

File: src/coprocessor/dag/expr/arithmetic.rs [hunk 8]
```diff
@@ -197,11 +197,94 @@ impl FnCall {
...
+        let lhs = try_opt!(self.children[0].eval_int(ctx, row));
```
Comment (AndreMouche) [inline src/coprocessor/dag/expr/arithmetic.rs]: check the divisor first like `tidb` did, it has better performance when the divisor is 0.

File: src/coprocessor/codec/overflow.rs [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+
+use coprocessor::dag::expr::{Error, Result};
+
+// div_i64 divides i64 a with b, returns:
+// - an Error indicating overflow occurred or the divisor is 0
+// - i64 otherwise
+#[inline]
+pub fn div_i64(a: i64, b: i64) -> Result<i64> {
+    if b == 0 {
+        Err(Error::division_by_zero())
```
Comment (AndreMouche) [inline src/coprocessor/codec/overflow.rs]: Could we use `return` to use  as less as possible `{}`

File: src/coprocessor/dag/expr/arithmetic.rs [hunk 10]
```diff
@@ -197,11 +197,94 @@ impl FnCall {
...
+        let res = match (lus, rus) {
+            (true, true) => Ok(((lhs as u64) / (rhs as u64)) as i64),
+            (false, false) => div_i64(lhs, rhs),
+            (false, true) => div_i64_with_u64(lhs, rhs as u64).map(|r| r as i64),
+            (true, false) => div_u64_with_i64(lhs as u64, rhs).map(|r| r as i64),
+        };
+        res.map(Some)
+    }
+
+    pub fn intdivide_decimal(&self, ctx: &mut EvalContext, row: &[Datum]) -> Result<Option<i64>> {
```
Comment (AndreMouche) [inline src/coprocessor/dag/expr/arithmetic.rs]: int_divide_decimal

File: src/coprocessor/dag/expr/arithmetic.rs [hunk 11]
```diff
@@ -197,11 +197,94 @@ impl FnCall {
...
+        let lhs = try_opt!(self.children[0].eval_decimal(ctx, row));
+        let rhs = try_opt!(self.children[1].eval_decimal(ctx, row));
+        let overflow = Error::overflow("BIGINT", &format!("({} DIV {})", lhs, rhs));
+        match lhs.into_owned() / rhs.into_owned() {
```
Comment (AndreMouche) [inline src/coprocessor/dag/expr/arithmetic.rs]: It seems we could use the function `divide_decimal` and convert the response to i64?

File: src/coprocessor/dag/expr/arithmetic.rs [hunk 12]
```diff
@@ -197,11 +197,94 @@ impl FnCall {
...
+        }
+    }
+
+    pub fn mod_int(&self, ctx: &mut EvalContext, row: &[Datum]) -> Result<Option<i64>> {
+        let lhs = try_opt!(self.children[0].eval_int(ctx, row));
+        let rhs = try_opt!(self.children[1].eval_int(ctx, row));
+        let lus = mysql::has_unsigned_flag(self.children[0].get_tp().get_flag());
+        let rus = mysql::has_unsigned_flag(self.children[1].get_tp().get_flag());
+        if rhs == 0 {
+            return Ok(None);
```
Comment (AndreMouche) [inline src/coprocessor/dag/expr/arithmetic.rs]: ditto, eval and check rhs first like `tidb` did.

**Comment by AndreMouche** [comment]
@zz-jason @XuHuaiyu PTAL

File: src/coprocessor/codec/overflow.rs (hunk 9 above)
Comment (bb7133) [reply src/coprocessor/codec/overflow.rs]: well, I thought this style(1 if-expression only in the body) would make the code more clear. If you prefer less '{}', I can change it

File: src/coprocessor/dag/expr/arithmetic.rs (hunk 8 above)
Comment (bb7133) [reply src/coprocessor/dag/expr/arithmetic.rs]: you're right, thanks!

File: src/coprocessor/dag/expr/fncall.rs [hunk 13]
```diff
@@ -218,7 +229,19 @@ impl FnCall {
             | ScalarFuncSig::InDecimal
             | ScalarFuncSig::InTime
             | ScalarFuncSig::InDuration
-            | ScalarFuncSig::InJson => (2, usize::MAX),
+            | ScalarFuncSig::InJson
+            | ScalarFuncSig::GreatestInt
```
[the remark this answers is not in the document]
Comment (bb7133) [reply src/coprocessor/dag/expr/fncall.rs]: okay

File: src/coprocessor/dag/expr/arithmetic.rs (hunk 12 above)
Comment (bb7133) [reply src/coprocessor/dag/expr/arithmetic.rs]: okay, thanks

File: src/coprocessor/dag/expr/arithmetic.rs (hunk 11 above)
Comment (bb7133) [reply src/coprocessor/dag/expr/arithmetic.rs]: I'm afraid we can't: divide_decimal and int_divide_decimal has different overflow error message('DECIMAL value is out of range in (xx / xx)' vs 'BIGINT value is out of range in (xx div xx)'. Future more,  we cannot convert this message through some regular expression, there is no way we can judge if an overflow error is returned by the division of divide_decimal(), or by the evaluation of the children(lhs/rhs)

**Comment by bb7133** [comment]
code updated, PTAL @AndreMouche @breeswish , thanks!

File: src/coprocessor/dag/expr/arithmetic.rs [hunk 14]
```diff
@@ -197,11 +197,96 @@ impl FnCall {
...
+            (true, false) => div_u64_with_i64(lhs as u64, rhs).map(|r| r as i64),
+        };
+        res.map(Some)
+    }
+
+    pub fn int_divide_decimal(&self, ctx: &mut EvalContext, row: &[Datum]) -> Result<Option<i64>> {
+        let rhs = try_opt!(self.children[1].eval_decimal(ctx, row));
+        let lhs = try_opt!(self.children[0].eval_decimal(ctx, row));
+        let overflow = Error::overflow("BIGINT", &format!("({} DIV {})", lhs, rhs));
+        match lhs.into_owned() / rhs.into_owned() {
```
[the remark this answers is not in the document]
Comment (bb7133) [reply src/coprocessor/dag/expr/arithmetic.rs]: It's not about the convert, I mean we cannot distinguish the overflows from nested divide_decimal().

But yes, we can extract some common codes and share them for divide_decimal/int_divide_decimal, I'll try it, thanks @AndreMouche 

By the way, in the related tidb code, some args are missing in ErrOverflow which caused bad error message, I can fix it too. @XuHuaiyu @zz-jason

**Comment by bb7133** [comment]
code updated @AndreMouche , thanks

**Comment by bb7133** [comment]
hi @XuHuaiyu PTAL, thanks!

**Comment by breeswish** [comment]
LGTM, @XuHuaiyu PTAL

**Comment by BusyJay** [comment]
Ping @XuHuaiyu

**Pull request merged**